design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,291 @@
+create and delete small clusters as needed using configuration supplied by the
+infrastructure owner (i.e. "clusters as cattle").
+
+#### Scenario 3: Infrastructure owner enables self-service, except for one on-prem cluster with sensitive data
+
+1. Infrastructure owner creates a variety of Kubernetes cluster classes with
+   configuration for different environments and different cloud providers.
+   Application owners are given permissions to create clusters using these
+   classes as they see fit.
+1. There is one particular set of workloads that is required to be run in an
+   existing on-premises cluster (for some important reason). The infrastructure
+   owner creates a Kubernetes `Secret` with connection information to that
+   cluster in an infrastructure namespace (secrets are namespace-scoped). They
+   then create a cluster-scoped Kubernetes `Provider` that references the
+   `Secret`.
+1. Application owner says that they need to run a sensitive workload on that
+   cluster. Infrastructure owner creates a new namespace called
+   `super-secret-workloads` and creates a `KubernetesTarget` in that namespace
+   that references the previously created Kubernetes `Provider`.
+1. Application owner provisions a `KubernetesApplication` with labels that
+   select that `KubernetesTarget` for provisioning.
+
+This example shows the power of this model, but also one of the notable
+shortcomings: if a user can publish clusters to a namespace with a
+`KubernetesTarget`, then they can consume any cluster. For the application
+owners that are doing self-service provisioning in this scenario, they will
+likely need to be able to create `KubernetesTarget` resources for their
+dynamically provisioned Kubernetes cluster managed resources. However, if they
+can do so, they could just as easily create a `KubernetesTarget` for the
+Kubernetes `Provider` that is only to be used for sensitive workloads. While it
+would be nice to restrict the clusters that could be published by users with the
+ability to create `KubernetesTarget` resources, there is not an immediately
+clear and elegant way to do so. It could be argued that only trusted users
+should have this permission level anyway, so it may be less of an issue than
+expressed here. Nevertheless, a potential solution is outlined
+[below](#future-considerations).
+
+## Technical Implementation
+
+The implementation of this proposal is relatively lightweight and requires
+minimal churn on any existing exposed APIs. The first step would be implementing
+the `KubernetesTarget` type in [core Crossplane]. In addition to the creation of
+the custom resource, a new controller would also be required to watch
+`KubernetesTarget` object creation and propagate `Secret` objects to its
+namespace as described above. This controller should use [predicates] to only
+reconcile `KubernetesTarget` objects that have *one* of either a `clusterRef`
+and `providerRef`. 
```

Comment:
I think that we should enforce one of either `clusterRef` or `providerRef` due to the fact that having one just take precedence over the other could lead to confusion about where a particular workload is being scheduled.

Reply:
I remember a similar discussion in the cross-resource referencing topic. It'd have been very helpful if OpenAPI v3 schema validation had _only one of these can be filled with a value_ type of validation.